### PR TITLE
Support manually-created and generated conversion functions

### PR DESCRIPTION
**Pull request by wojtek-t**
Part of #6800

This was mentioned in https://github.com/GoogleCloudPlatform/kubernetes/pull/7556#issuecomment-98969053
Motivated by #7814, where a lot of unnecessary boilerplate had to be written.

cc @smarterclayton @lavalamp @fgrzadkowski

### CODE REVIEW

**Comment by smarterclayton** [comment]
Test case that proves a generated conversion is overriden?

**Comment by wojtek-t** [comment]
>
Test case that proves a generated conversion is overriden?

Good point - test added.
PTAL

**Comment by mbforbes** [comment]
Without digging into the code, I'm assuming the motivation sprung from #7343 (of which #7814 is the un-revert), and so @erictune and @pmorie are good candidates for review. Assigning to @pmorie for now; please feel free to shuffle around.

**Comment by smarterclayton** [comment]
I'm going to grab this

**Comment by smarterclayton** [comment]
LGTM

**Comment by wojtek-t** [comment]
Thanks @smarterclayton !

I think that it would be good for @lavalamp to also take a look into it.

File: pkg/conversion/converter.go [hunk 1]
```diff
@@ -278,10 +268,47 @@ func (c *Converter) RegisterConversionFunc(conversionFunc interface{}) error {
...
+	return nil
+}
+
+// RegisterConversionFunc registers a conversion func with the
+// Converter. conversionFunc must take three parameters: a pointer to the input
+// type, a pointer to the output type, and a conversion.Scope (which should be
+// used if recursive conversion calls are desired).  It must return an error.
+//
+// Example:
+// c.RegisteConversionFunc(
```
Comment (lavalamp) [inline pkg/conversion/converter.go]: nit: missing r

**Comment by lavalamp** [comment]
Thanks for fixing this!

File: pkg/conversion/converter.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/conversion/converter.go]: Will fix that in a subsequenct PR

**Comment by thockin** [comment]
Is there somewhere I can read an overview of this changeset?  I was very familiar with the state before autogeneration, and I want to understand what this is buying us and how I interact with it for manual conversion functions...

**Comment by lavalamp** [comment]
@thockin see the list of PRs on the end of #6800. This is an... expedient... way to cut the reflection out of the critical path so we can meet our performance goals. I think we can probably come up with a more user friendly mechanism but that's going to have to be post 1.0.

**Comment by lavalamp** [comment]
The overview is that we're autogenerating conversion functions--while allowing manual override--instead of using the conversion logic.

**Comment by thockin** [comment]
As an occasional author of manual conversion functions, what has changed
for me?

**Comment by lavalamp** [comment]
@wojtek-t can confirm (and we'll want to write this up really soon), but I believe the steps will be:

1. Move the auto-generated function from the list in `newer.Scheme.AddGeneratedConversionFuncs(...)` to the list in `newer.Scheme.AddConversionFuncs(...)` (soon: move the body of the function out of the file containing the auto-generated functions)
2. Edit it.
3. Regenerate the conversion functions.

**Comment by smarterclayton** [comment]
I did this today for the etcd portal ip allocator and was easy (not quite fully documented yet).

**Comment by wojtek-t** [comment]
Yes - this is exactly what @lavalamp wrote above. There is one more significant PR #7556 that will come (it's already under review) and as soon as it's in I will update documentation.
I already kind of updated it at some point (docs/devel/api_changes.md), but I agree it's again updated :)

**Pull request merged**